### PR TITLE
Added event marks on dates

### PR DESCRIPTION
**Pull request by MyMaloks**
Allow to add colored marks below the date in the table cells as shown in the example
jsCalendar

### CODE REVIEW

**Comment by perskes** [comment]
What a coincidence, I was lookin gin the docs for that feature and wanted to make a feature request, as I could not find it!
This is great and I'd love to see this in jsCalendar!

Basically any "Event Mark" would be nice, but your approach is multifunctional, I like it!

**Comment by GramThanos** [comment]
I would like to include such features as extensions, like the datepicker extension, so that people can load them only if they need them. So, this request will have to wait a little (so that the current beta version is released and the documentation is updated with all the API)

**Comment by perskes** [comment]
Hey GramThanos, 
thanks for the reply.

It sounds like a good idea, until then I will use MyMaloks fork.
Looking forward for further updates and improvements!

Thank you!

**Comment by MyMaloks** [comment]
@perskes , thanks for the your interest in the feature I'd like to introduce.
Please be aware that it is currently only possible to add event marks (no clear all marks or remove specific colored marks).

I will however probably update this soon.
@GramThanos , I actually did not yet see how you impemented extensions. I may update it to fit your requirements

**Comment by GramThanos** [comment]
@MyMaloks yes I know, it's a little messy right now (the hole version-documentation stuff)

File: examples/calendar marked events.html [hunk 1]
```diff
@@ -0,0 +1,77 @@
+<!DOCTYPE html>
+<html lang="en">
+	<head>
+		<meta charset="utf-8">
+		<title>jsCalendar</title>
+		<meta name="description" content="jsCalendar example">
+		<meta name="author" content="GramThanos">
```
Comment (GramThanos) [inline examples/calendar marked events.html]: Change author

File: source/jsCalendar.css [hunk 2]
```diff
@@ -130,6 +130,27 @@
...
+			display: inline;
+			margin: 0px 1px;
+		}
+		.jsCalendar tbody td div.jsCalendar-event-marks .jsCalendar-event-more-mark {
+			font-size: 8px;
+			position: relative;
+			bottom: 16px;
+			right: -2px;
+		}
+
```
Comment (GramThanos) [inline source/jsCalendar.css]: Add these rules to a separate file, maybe named `jsCalendar.eventmarks.css`

File: source/jsCalendar.js [hunk 3]
```diff
@@ -97,6 +97,7 @@ var jsCalendar = (function(){
         this._now = null;
```
Comment (GramThanos) [inline source/jsCalendar.js]: Move the code to separate extension file, maybe named `jsCalendar.eventmarks.js`

Based on your code, extension code :

```javascript
(function($){
	// Check if jsCalendar exists
	if (!$) throw 'Error jsCalendar library was not found';

	// Example Extension
	var init = function(instance) {
		instance._eventMarks = {};
	};
	var update = function(instance, month) {
		for (var i = month.days.length - 1; i >= 0; i--) {
			// If date is marked for events
			var cellContent = instance._elements.bodyCols[i].textContent;
			if (month.days[i].getTime() in instance._eventMarks) {
				cellContent += '<div class="jsCalendar-event-marks">'
				var moreColors = 0;
				var colors = instance._eventMarks[month.days[i].getTime()];
				for (var j = 0; j < colors.length; j++) {
					if (j < 3) {
						cellContent += '<div class="jsCalendar-event-mark" style="background-color: ' + colors[j] + '"></div>';
					}
					else moreColors++;
				}
				if (moreColors > 0) {
					cellContent += '<div class="jsCalendar-event-more-mark">+' + moreColors + '</div>';
				}
				cellContent += '</div>';
				instance._elements.bodyCols[i].innerHTML = cellContent;
			}
		}
	};

	$.prototype._addEventMarksForDate = function(date, eventColors) {
		date = $.tools.parseDate(date);
		date.setHours(0, 0, 0, 0);
		date = date.getTime();

		if (!(date in this._eventMarks)) {
			this._eventMarks[date] = [];
		}

		this._eventMarks[date] = this._eventMarks[date].concat(eventColors);
	};

	$.prototype._removeEventMarksForDate = function(date, eventColors) {
		date = $.tools.parseDate(date);
		date.setHours(0, 0, 0, 0);
		date = date.getTime();

		if (date in this._eventMarks) {
			for (var i = 0; i < eventColors.length; i++) {
				var pos = this._eventMarks[date].indexOf(eventColors[i]);
				if (pos >= 0) {
					this._eventMarks[date].splice(pos, 1);
				}
			}
		}
	};

	$.prototype._clearEventMarksForAllDates = function() {
		this._eventMarks = {};
	};

	// Select dates
	$.prototype.addEventMark = function(date, eventColors){
		// If no arguments
		if (typeof date === 'undefined') {
			// Return
			return this;
		}

		// If eventColors not array
		if (!(eventColors instanceof Array)) {
			// Lets make it an array
			eventColors = [eventColors];
		}
		// Mark dates
		this._addEventMarksForDate(date, eventColors);
		// Refresh
		this.refresh();

		// Return
		return this;
	};

	// Select dates
	$.prototype.removeEventMark = function(date, eventColors){
		// If no arguments
		if (typeof date === 'undefined') {
			// Return
			return this;
		}

		// If eventColors not array
		if (!(eventColors instanceof Array)) {
			// Lets make it an array
			eventColors = [eventColors];
		}
		// Unmark dates
		this._removeEventMarksForDate(date, eventColors);
		// Refresh
		this.refresh();

		// Return
		return this;
	};

	// Clear event marks from all dates
	$.prototype.clearAllEventMarks = function(){
		// Unselect all dates
		this._clearEventMarksForAllDates();
		// Refresh
		this.refresh();

		// Return
		return this;
	};

	// Load extension
	$.ext('EventMarks', {
		init : init,
		update : update
	});

})(window.jsCalendar);
```

**Review by GramThanos — CHANGES_REQUESTED** [review]
Sorry for the long delay. So I added the extension functionality that I was planning to add. I ported your code to an extension (code in comments).

I suggest moving your code to 3 files:
`examples/extensions/calendar marked events.html`
`extensions/jsCalendar.eventmarks.js`
`extensions/jsCalendar.eventmarks.css`

If you want help comment bellow, this is the first time I am using the GitHub's review system, so excuse me if I make any mistake while reviewing.

**Comment by GramThanos** [comment]
Updated example:
image

**Pull request merged**